src/modules/royalty.ts

```diff
@@ -0,0 +1,203 @@
+  }
+
+  /**
+   * @internal
+   */
+  protected getModuleType(): ModuleType {
+    return SplitsModule.moduleType;
+  }
+
+  public async get(): Promise<Currency> {
+    return await getCurrencyMetadata(this.providerOrSigner, this.address);
+  }
+
+  public async getAllRecipients(): Promise<SplitRecipient[]> {
+    const recipients: SplitRecipient[] = [];
+
+    let index = BigNumber.from(0);
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
+      try {
+        const recipientAddress = await this.readOnlyContract.payee(index);
+        recipients.push(
+          await this.getRecipientSplitPercentage(recipientAddress),
+        );
+        index = index.add(1);
+      } catch (err: any) {
+        // The only way we know how to detect that we've found all recipients
+        // is if we get an error when trying to get the next recipient.
+        if (
+          "method" in err &&
+          (err["method"] as string).toLowerCase().includes("payee(uint256)")
+        ) {
+          break;
+        } else {
+          throw err;
+        }
+      }
+    }
+
+    return recipients;
+  }
+
+  public async getRecipientSplitPercentage(
+    address: string,
+  ): Promise<SplitRecipient> {
+    const [totalShares, walletsShares] = await Promise.all([
+      this.readOnlyContract.totalShares(),
+      this.readOnlyContract.shares(address),
+    ]);
+
+    return {
+      address,
+      splitPercentage:
+        walletsShares.mul(BigNumber.from(1e7)).div(totalShares).toNumber() /
+        1e5,
+    };
+  }
+
+  public async balanceOf(address: string): Promise<BigNumber> {
+    const walletBalance = await this.readOnlyContract.provider.getBalance(
+      this.address,
+    );
+    const totalReleased = await this.readOnlyContract["totalReleased()"]();
+    const totalReceived = walletBalance.add(totalReleased);
+
+    return this._pendingPayment(
+      address,
+      totalReceived,
+      await this.readOnlyContract["released(address)"](address),
+    );
+  }
+
+  public async balanceOfByToken(
+    walletAddress: string,
+    tokenAddress: string,
+  ): Promise<CurrencyValue> {
+    const erc20 = ERC20__factory.connect(tokenAddress, this.providerOrSigner);
+    const walletBalance = await erc20.balanceOf(this.address);
+    const totalReleased = await this.readOnlyContract["totalReleased(address)"](
+      tokenAddress,
+    );
+    const totalReceived = walletBalance.add(totalReleased);
+    const value = await this._pendingPayment(
+      walletAddress,
+      totalReceived,
+      await this.readOnlyContract["released(address,address)"](
+        tokenAddress,
+        walletAddress,
+      ),
+    );
+    return await getCurrencyValue(this.providerOrSigner, tokenAddress, value);
+  }
+
+  public async withdraw(address: string): Promise<void> {
+    await this.sendTransaction("release(address)", [address]);
+  }
+
+  private async _pendingPayment(
```

Comment:
I would *like* it if private & public methods were not mixed together ... but this is truly just personal preference